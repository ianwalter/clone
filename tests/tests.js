const { test } = require('@ianwalter/bff')
const Vue = require('vue')
const Vuex = require('vuex')
const clone = require('..')

Vue.use(Vuex)

test`an array is cloned ${t => {
  const arr = [{ mutated: false }]
  const copy = clone(arr)
  t.expect(copy).toStrictEqual(arr)
  arr.push('mutated')
  arr[0].mutated = true
  t.expect(copy).toMatchSnapshot()
}}`

test`a map is cloned ${t => {
  const map = new Map()
  map.set('one', { mutated: false })
  const copy = clone(map)
  t.expect(copy).toStrictEqual(map)
  const one = map.get('one')
  one.mutated = true
  map.set('two', 'mutated')
  t.expect(copy).toMatchSnapshot()
}}`

test`a nested array is cloned ${t => {
  const nested = { arr: [{ mutated: false }] }
  const copy = clone(nested)
  t.expect(copy).toStrictEqual(nested)
  nested.arr.push('mutated')
  nested.arr[0].mutated = true
  t.expect(copy).toMatchSnapshot()
}}`

test`Vuex store state prototype properties are not copied ${t => {
  const message = 'Hello!'
  const name = 'The Wheel'
  const store = new Vuex.Store({
    state: { message },
    modules: {
      song: {
        namespaced: true,
        state: { name },
        mutations: {
          name: (state, name) => (state.name = name)
        }
      }
    }
  })
  const copy = clone(store.state)
  t.expect(copy).toMatchSnapshot()
  store.commit('song/name', 'Desire')
  t.expect(copy.song.name).toBe(name)
}}`

test`circular properties are not copied ${t => {
  function Podcast () {
    this.name = 'Beanicles'
    this.circular = this
  }
  const podcast = new Podcast()
  const copy = clone(podcast)
  t.expect(copy).toEqual({ name: podcast.name })
}}`
