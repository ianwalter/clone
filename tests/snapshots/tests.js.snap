// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vuex store state prototype properties are not copied 1`] = `
Object {
  "message": "Hello!",
  "song": Object {
    "name": "The Wheel",
  },
}
`;

exports[`a map is cloned 1`] = `
Map {
  "one" => Object {
    "mutated": false,
  },
}
`;

exports[`a nested array is cloned 1`] = `
Object {
  "arr": Array [
    Object {
      "mutated": false,
    },
  ],
}
`;

exports[`an array is cloned 1`] = `
Array [
  Object {
    "mutated": false,
  },
]
`;
